enhancements/openshift-console-cloud-shell.md

```diff
@@ -0,0 +1,195 @@
+
+It is secured using a JWT proxy sidecar that verifies that the user is authorized to access the terminal. A user is only authorized to access his workspaces terminals.
+
+### Implementation details
+
+Eclipse Che workspaces already provide a web based secured shell. The idea is to reuse part of a Che workpace components to implement the OpenShift Console Cloud Shell. In the following sections we are going into the details of:
+- How does a Cloud Shell instance compares to an Eclipse Che workspace
+- Authentication and authorization mechanism to secure CloudShell instances
+- The mechanism to deny any user other then the owner to `exec` into a Cloud Shell Pod
+
+#### Cloud Shell vs Che Workspace
+
+A cloud shell instance will be implemented as a minimal Che workspace pod that includes the following 3 containers:
+
+- [OpenShift OAuth Proxy](https://github.com/openshift/oauth-proxy): for authentication
+- Terminal server: connects to `xterm.js` and does `exec` to spawn shells in the tooling container 
+- Tooling container: with CLI tools like `bash`, `oc`, `odo`, `jq`, `vim` and stuff
+
+![](https://i.imgur.com/3mSAe1w.png)
+
+
+Such a workspace will be defined using a `DevWorkspace` Custom Resource (c.f. [#15425](https://github.com/eclipse/che/issues/15425)):
+
+```yaml
+apiVersion: workspaces.ecd.eclipse.org/v1alpha1
+kind: DevWorkspace
+metadata:
+  name: cloudshell
+spec:
+  components:
+    - type: dockerimage
+      id: eclipse/cloudshell/latest
+``` 
+
+As a consequence, the Cloud Shell will be dependent on the Eclipse Che DevWorkspace operator that deploys the CRD and controller to manage workspaces. But unlike Eclipse Che, in a Cloud Shell instance:
+- Eclipse Che server components are not deployed: `wsmaster`, `registries`, `keycloak` and `postgres`. Instanciation of `DevWorkspaces` is a responsibility of the OpenShift Console.
+- Eclipse Che and the Cloud Shell are both deployed and controlled using operators but those are distinct ones.
+- Theia (or other editors) based workspaces are not supported. Only Eclipse Che workspaces with editor of type CloudShell are supported.
+
+#### Authentication and Authorization
+
+The Terminal server endpoint is exposed through a TLS route. That's the endpoint that the clients running on the browser side (xterm.js) will connect to.
+
+![](https://i.imgur.com/XWcgHeI.png)
+
+The endpoint is secured through an [OpenShift OAuth Proxy](https://github.com/openshift/oauth-proxy) sidecar. The client includes the user OpenShift token in its requests HTTP header. Only authenticated OpenShift users that satisfy the RBAC policy requirements are able to send requests to the Terminal server.
```

Comment:
if you have the ability to indicate that only userX can exec or attach into PodA, why do you need to separately authenticate them.  Can't you just use a normal exec call?

Reply:
Currently, the web terminal we use has two API methods:
- connect - initialize exec to k8s API
- attach - connect to initialized exec.

This design was made from the beginning on Che and it allows (but not sure if it's implemented, maybe for Theia) to restore the terminal state after user refreshes a page.

But as an alternative - we could leave the only one method which would do both operations, it would allow avoiding the need to protect our endpoints with authentication. 

But then the client should be somehow able to get k8s client.
In case, of OpenShift OAuth Proxy - it does it for CloudShell: use OpenShiftOAuth to get scoped token, store it in the cookies, and it let CloudShell client just send a request and authorization header will be provided by browser.

So, I hope it makes the current implementation a bit cleaner but we're able to discuss and find other possible implementations.